core/src/main/java/hudson/slaves/NodeProvisioner.java

```diff
@@ -217,29 +218,40 @@ public void run() {
                         PlannedNode f = itr.next();
                         if (f.future.isDone()) {
                             try {
-                                Node node = f.future.get();
-                                for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
-                                    cl.onComplete(f, node);
-                                }
-    
-                                jenkins.addNode(node);
-                                LOGGER.log(Level.INFO,
-                                        "{0} provisioning successfully completed. " 
-                                                + "We have now {1,number,integer} computer(s)",
-                                        new Object[]{f.displayName, jenkins.getComputers().length});
-                            } catch (InterruptedException e) {
-                                throw new AssertionError(e); // since we confirmed that the future is already done
-                            } catch (ExecutionException e) {
-                                LOGGER.log(Level.WARNING, "Provisioned agent " + f.displayName + " failed to launch",
-                                        e.getCause());
-                                for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
-                                    cl.onFailure(f, e.getCause());
+                                Node node = null;
+                                try {
+                                    node = f.future.get();
+                                    for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
```

Comment:
I would move this outside of the try catch. That way we do not risk loss of a `Node` instance if a `CloudProvisioningListener` happens to propagate an `ExecutionException`...